jiant/proj/main/metarunner.py

```diff
@@ -122,6 +125,27 @@ def save_model(self):
             file_name=f"model__{self.train_state.global_steps:09d}",
         )
 
+    def save_last_model_with_metadata(self):
+        save_model_with_metadata(
+            model=self.model,
+            metadata={
+                "train_state": self.train_state.to_dict(),
+            },
+            output_dir=self.output_dir,
+            file_name="last_model",
+        )
+
+    def save_best_model_with_metadata(self, val_metrics_dict):
+        save_state_dict_with_metadata(
+            state_dict=self.best_state_dict,
+            metadata={
+                "val_state": self.best_val_state.to_dict(),
+                "val_metrics": self.best_val_state.metrics,
+            },
+            output_dir=self.output_dir,
+            file_name="best_model",
+        )
+
```

Comment:
These functions seem like they are doing the same thing in different ways. Can they be combined?

Reply:
This also looks like a potential case of excessive delegation.

Reply:
It looks excessive, but the specific reason for this is to allow these methods to be overridden. (E.g. for adapters, we don't want to save all the model's weights, so we need a custom save function). Happy to implement another way, but I've found this helpful.

Reply:
Side benefit: can override for testing

Reply:
I still think there is room for refactoring here. The function signature should similar to:
```python
def save_model(self, model, file_name, metadata_dict)
```